# Linear algebra with Numpy 🧠🧠

Some practice exercises are available the [Linear Algebra prepwork](https://app.jedha.co/course/linear-algebra-basics/linear-algebra-exercises).
In this exercise, you will apply what you learnt in the "Numpy and Linear Algebra" to check numerically the answers to these exercises.

In [1]:
import numpy as np

## Operations with vectors and matrices 🤹‍♂️🤹‍♂️

### Vector operations

1. Compute the sum $\vec{u} + \vec{v}$ of the following vectors :
$$
\vec{u} =
\begin{bmatrix}
2 \\
9 \\
-1
\end{bmatrix}
$$
$$
\vec{v} =
\begin{bmatrix}
1 \\
5 \\
3
\end{bmatrix}
$$

In [3]:
u_vec = np.array([2,9,-1], dtype=int)
v_vec = np.array([1,5,3], dtype=int)

u_vec + v_vec

array([ 3, 14,  2])

2. Compute the difference $\vec{u} - \vec{v}$ between the following vectors :
$$
\vec{u} =
\begin{bmatrix}
10 \\
-2 \\
1
\end{bmatrix}
$$
$$
\vec{v} =
\begin{bmatrix}
4 \\
3 \\
2
\end{bmatrix}
$$

In [4]:
u_vec = np.array([10,-2,1], dtype=int)
v_vec = np.array([4,3,2], dtype=int)

u_vec - v_vec

array([ 6, -5, -1])

3. Compute : $\alpha (\vec{u} + \vec{v})$ with $\alpha = 4$ and :
$$
\vec{u} =
\begin{bmatrix}
1 \\
-2 \\
2
\end{bmatrix}
$$
$$
\vec{v} =
\begin{bmatrix}
2 \\
1 \\
0
\end{bmatrix}
$$

In [5]:
alpha = 4

u_vec = np.array([1,-2,2])
v_vec = np.array([2,1,0])

alpha*(u_vec + v_vec)

array([12, -4,  8])

4. Compute the norms $\Vert\vec{u}\Vert$ and $\Vert\vec{v}\Vert$ with:
$$
\vec{u}
=
\begin{bmatrix}
1 \\
2 \\
3 \\
4 \\
\end{bmatrix}
$$
$$
\vec{v}
=
\begin{bmatrix}
-1 \\
2 \\
-3 \\
4 \\
\end{bmatrix}
$$

In [6]:
u_vec = np.array([1,2,3,4])
v_vec = np.array([-1,2,-3,4])

print(f"Norm u: {np.linalg.norm(u_vec)}")
print(f"Norm v: {np.linalg.norm(v_vec)}")

Norm u: 5.477225575051661
Norm v: 5.477225575051661


### Matrix operations

5. Compute the sum $A + B$ of the two following matrices :
$$
A =
\begin{bmatrix}
1 & 2 & 3 \\
2 & 1 & 4 \\
3 & 4 & 1
\end{bmatrix}
$$
$$
B =
\begin{bmatrix}
2 & 3 & 4 \\
3 & 2 & 5 \\
4 & 5 & 2
\end{bmatrix}
$$

In [7]:
A_mat = np.array([[1, 2, 3],
                  [2, 1, 4],
                  [3, 4, 1]])

B_mat = np.array([[2, 3, 4],
                  [3, 2, 5],
                  [4, 5, 2]])

A_mat + B_mat

array([[3, 5, 7],
       [5, 3, 9],
       [7, 9, 3]])

6. Compute the operation $2A - B$ with:
$$
A =
\begin{bmatrix}
1 & 2 & 3 \\
2 & 1 & 4 \\
3 & 4 & 1
\end{bmatrix}
$$
$$
B =
\begin{bmatrix}
2 & 3 & 4 \\
3 & 2 & 5 \\
4 & 5 & 2
\end{bmatrix}
$$

In [8]:
A_mat = np.array([[1, 2, 3],
                  [2, 1, 4],
                  [3, 4, 1]])

B_mat = np.array([[2, 3, 4],
                  [3, 2, 5],
                  [4, 5, 2]])

2*A_mat - B_mat

array([[0, 1, 2],
       [1, 0, 3],
       [2, 3, 0]])

7. Compute the products of matrices : $AB$ and $BA$ with:
$$
A =
\begin{bmatrix}
1 & 0 & 2 \\
2 & 1 & 0 \\
1 & 2 & 1
\end{bmatrix}
$$
$$
B =
\begin{bmatrix}
2 & 0 & 0 \\
1 & 2 & 0 \\
0 & 1 & 2
\end{bmatrix}
$$

In [12]:
A_mat = np.array([[1, 0, 2],
                  [2, 1, 0],
                  [1, 2, 1]])

B_mat = np.array([[2, 0, 0],
                  [1, 2, 0],
                  [0, 1, 2]])

print("AB:")
print(A_mat @ B_mat)
print(" ")
print("BA:")
print(B_mat.dot(A_mat))

AB:
[[2 2 4]
 [5 2 0]
 [4 5 2]]
 
BA:
[[2 0 4]
 [5 2 2]
 [4 5 2]]


8. Compute the products of matrices : $AB$ and $BA$ with:
$$
A =
\begin{bmatrix}
1 & 0 \\
2 & 1 \\
1 & 2
\end{bmatrix}
$$
$$
B =
\begin{bmatrix}
2 & 0 & 0 \\
1 & 2 & 0
\end{bmatrix}
$$

In [13]:
A_mat = np.array([[1, 0],
                  [2, 1],
                  [1, 2]])

B_mat = np.array([[2, 0, 0],
                  [1, 2, 0]])

print("AB:")
print(A_mat @ B_mat)
print(" ")
print("BA:")
print(B_mat @ A_mat)

AB:
[[2 0 0]
 [5 2 0]
 [4 4 0]]
 
BA:
[[2 0]
 [5 2]]


### Matrix-vector product

9. Compute the matrix-vector product $A\vec{v}$ with :
$$
A =
\begin{bmatrix}
1 & 0 & 2 \\
2 & 1 & 0 \\
1 & 2 & 1
\end{bmatrix}
$$
$$
\vec{v} =
\begin{bmatrix}
-1 \\
1 \\
2
\end{bmatrix}
$$

In [15]:
A_mat = np.array([[1, 0, 2],
                  [2, 1, 0],
                  [1, 2, 1]])

v_vec = np.array([-1, 1, 2])

print(A_mat @ v_vec)

[ 3 -1  3]


## Collinearity and orthogonality 🏌️🏌️

10. Compute the dot product $\vec{u} \cdot \vec{v} $ with:
$$
\vec{u}
=
\begin{bmatrix}
2 \\
-3 \\
-1 \\
\end{bmatrix}
$$
$$
\vec{v}
=
\begin{bmatrix}
3 \\
1 \\
3 \\
\end{bmatrix}
$$

What can you conclude about the angle between the two vectors ?

In [ ]:
u_vec = np.array([2, -3, -1])
v_vec = np.array([3, 1, 3])

print(f"dot product: {u_vec.dot(v_vec)}")

# on peut conlure que si le produit des 2 vecteurs est égale à 0, ils sont orthogonaux. theta = 90°

dot product: 0


11. Are $\vec{u}$, $\vec{v}$ collinear ?
$$
\vec{u}
=
\begin{bmatrix}
4 \\
2 \\
-1
\end{bmatrix}
$$

$$
\vec{v}
=
\begin{bmatrix}
-1 \\
-\frac{1}{2} \\
\frac{1}{4}
\end{bmatrix}
$$

In [ ]:
u_vec = np.array([4, 2, -1])

v_vec = np.array([-1, -1/2, 1/4])

print(f"Element-wise division: {u_vec / v_vec}")

# Oui ils sont colinéaires, facteur -4, u = -4 v

Element-wise division: [-4. -4. -4.]


12. Are $\vec{u}$, $\vec{v}$ and $\vec{w}$ linearly independent (i.e. there's no relationship of multicolinearity between these vectors) ?
$$
\vec{u}
=
\begin{bmatrix}
1 \\
2
\end{bmatrix}
$$

$$
\vec{v}
=
\begin{bmatrix}
-2 \\
1 \\
\end{bmatrix}
$$

$$
\vec{w}
=
\begin{bmatrix}
-1 \\
3
\end{bmatrix}
$$

In [ ]:
u_vec = np.array([1, 2])
v_vec = np.array([-2, 1])
w_vec = np.array([-1, 3])

print(f"u: {u_vec}")
print(f"w - v: {w_vec - v_vec}")

# We just showed that u = w − v which means the vectors are not linearly independent !

u: [1 2]
w - v: [1 2]


## Systems of linear equations 🧩🧩

Consider the following system of linear equations :

$$
\begin{equation}
    \begin{cases}
      2x_1 + x_2 = 2\\
      x_1 + 2x_2 + x_3 = 0 \\
      x_2 + 2x_3 = 1
    \end{cases}
\end{equation}
$$

13. We'd like to write the matrix representation of this system of equations, in the form $A\vec{x} = \vec{c}$ where :
- $A$ is a matrix containing the coefficients of the equation
- $\vec{x}$ is the vector of unknowns
- $\vec{c}$ is a vector of constant values

Create a numpy array corresponding to $A$ and another corresponding to $\vec c$

In [22]:
A_mat = np.array([[2, 1, 0],
                  [1, 2, 1],
                  [0, 1, 2]])

c_vec = np.array([2, 0, 1])

14. Use numpy to solve the system of equations

In [23]:
# methode 1 avec la matrice inverse
print("-- Method 1 --")
A_mat_inv = np.linalg.inv(A_mat)
# resolution equation
x_vec = A_mat_inv @ c_vec
print("Solution of equation Ax = c:")
print("x = ", x_vec)
print()

# methode 2: with np.linalg.solve
print("-- Method 2 --")
x_vec = np.linalg.solve(A_mat, c_vec)
print("Solution of equation Ax = c:")
print("x = ", x_vec)

-- Method 1 --
Solution of equation Ax = c:
x =  [ 1.75 -1.5   1.25]

-- Method 2 --
Solution of equation Ax = c:
x =  [ 1.75 -1.5   1.25]


15. **[Bonus question - for experts only 💪]** Let's use the concept of matrix inverse to solve the system of linear equations !

* Prove that the following matrix is the inverse of matrix $A$ :
$$
B = \begin{bmatrix}
\frac{3}{4} & - \frac{1}{2} & \frac{1}{4} \\
-\frac{1}{2} & 1 & -\frac{1}{2} \\
\frac{1}{4} & -\frac{1}{2} & \frac{3}{4}
\end{bmatrix}
$$

* And then, use $A^{-1}$ to find the solution to the system of equation

In [26]:
B_mat = np.array([[3/4, -1/2, 1/4],
                  [-1/2, 1, -1/2],
                  [1/4, -1/2, 3/4]])

B_mat_inv = np.linalg.inv(B_mat)
print(f"Inverse of B: {B_mat_inv}")
print(f"Inverse of A: {A_mat_inv}")

x_vec = A_mat_inv @ c_vec
print("Solution of equation Ax = c:")
print(f"Solution: x = {x_vec}")

Inverse of B: [[2. 1. 0.]
 [1. 2. 1.]
 [0. 1. 2.]]
Inverse of A: [[ 0.75 -0.5   0.25]
 [-0.5   1.   -0.5 ]
 [ 0.25 -0.5   0.75]]
Solution of equation Ax = c:
Solution: x = [ 1.75 -1.5   1.25]


## Matrix diagonalization 🤸🤸

15. Consider the following matrix :

$$
A =
\begin{bmatrix}
3 & -1 \\
2 & 0
\end{bmatrix}
$$

a. Show that $\vec{u} = \begin{bmatrix}1\\1\end{bmatrix}$ and $\vec{v} = \begin{bmatrix}1\\2\end{bmatrix}$ are eigenvectors of $A$

b. What are the eigenvalues associated to $\vec{u}$ and $\vec{v}$ ?

In [33]:
A_mat = np.array([[3, -1],
                  [2, 0]])

eigenvals, eigenvecs = np.linalg.eig(A_mat)
# Method 2 - with np.linalg.eig
print("eigenvectors of A:")
print(eigenvecs)
print("eigenvalues of A:")
print(eigenvals)

eigenvectors of A:
[[0.70710678 0.4472136 ]
 [0.70710678 0.89442719]]
eigenvalues of A:
[2. 1.]


We can notice that:
* $A \vec u = 2 \vec u$ which means that $\vec u$ is an eigenvector of $A$ with eigenvalue 2
* $A \vec v = \vec v$ which means that $\vec u$ is an eigenvector of $A$ with eigenvalue 1.

In [34]:
# Method 1 - compute Au and Av

A = np.array([[3, -1],
              [2, 0]])

u = np.array([1, 1])
v = np.array([1, 2])

print("Au: ", A @ u)
print("Av: ", A @ v)


Au:  [2 2]
Av:  [1 2]


💡With this second method, we find the same eigenvalues but not the same components for the eigenvectors. This is because np.linalg.eig finds normalized eigenvectors (= the norm is equal to 1). Our vectors u and v are actually colinear to the eigenvectors found by np.linalg.eig:

- 0.707u=[0.707 | 0.707]
- 0.447v=[0.447 | 0.894]